### PR TITLE
gh-95205: Improve wasm README

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-95205

### CODE REVIEW

**Comment by erlend-aasland** [comment]
cc. @smontanaro

File: Tools/wasm/README.md
Comment (erlend-aasland) [inline Tools/wasm/README.md]:
```suggestion
From within your local CPython repo clone, run one of the following commands:
```

File: Tools/wasm/README.md
Comment (erlend-aasland) [inline Tools/wasm/README.md]: Christian: IMO, we should remove the `-C` here. You do not necessarily want to reuse the configure cache from your host inside the container.

**Comment by smontanaro** [comment]
LGTM. I think the `-C` flag should go as well.

File: Tools/wasm/README.md
Comment (erlend-aasland) [reply Tools/wasm/README.md]:
```suggestion
../../cpython/configure
```

**Comment by erlend-aasland** [comment]
> LGTM. I think the `-C` flag should go as well.

Thanks for reviewing. I've added the removal of `-C` as a suggestion; waiting for Christian to chime in before I land this.

**Comment by smontanaro** [comment]
Note that there are missing "cpython" references in the "Cross compile to wasm32-emscripten for browser" section as well.

Finally (I think)... For Docker naifs it should be made explicit that when exiting and then reentering the docker world, previous work will be lost. If possible, it might well be worth showing users how to use Christian's Docker image as the base for creating a build environment which is preserved across runs. If I'm not making myself clear, let me know and I'll try to expand a bit.

**Comment by erlend-aasland** [comment]
PTAL.

> If possible, it might well be worth showing users how to use Christian's Docker image as the base for creating a build environment which is preserved across runs.

I'm not sure how explicit we should be regarding this. Let me know if my last update was too vague regarding this.

File: Tools/wasm/README.md
Comment (erlend-aasland) [inline Tools/wasm/README.md]:
```suggestion
  emconfigure ../../cpython/configure \
```

File: Tools/wasm/README.md
Comment (erlend-aasland) [inline Tools/wasm/README.md]:
```suggestion
  emconfigure ../../cpython/configure \
```

File: Tools/wasm/README.md
Comment (erlend-aasland) [reply Tools/wasm/README.md]: I pushed this in 330ca5ef4ac798d057b53210fdb9907f70040d76

**Comment by tiran** [comment]
Leave the ``config.cache`` in the instructions. It is a massive time safer for everybody that wants to hack on WASM. ``emconfigure configure`` takes several minutes even on a fast machine.

File: Tools/wasm/README.md
Comment (tiran) [inline Tools/wasm/README.md]: The old path was correct. The new path won't work.

Comment (erlend-aasland) [reply Tools/wasm/README.md]: No, I don't think it is so; the old path is wrong.

~But let's be more explicit and use the absolute path from within the container.~ OTOH; no need, since the new path is correct.

Comment (erlend-aasland) [reply Tools/wasm/README.md]:
```console
$ docker run --rm -ti -v $(pwd):/python-wasm/cpython quay.io/tiran/cpythonbuild:emsdk3
Setting environment variables:
PATH = /emsdk:/emsdk/upstream/emscripten:/emsdk/node/14.18.2_64bit/bin:/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
root@27c144c9747d:/python-wasm# pwd
/python-wasm
root@27c144c9747d:/python-wasm# ls
README.md   build-python-build.sh               build-python-emscripten-node.sh  clean-host.sh  fetch-python.sh        run-python-node.sh  test-emscripten-node.sh
browser-ui  build-python-emscripten-browser.sh  build-python-wasi.sh             cpython        run-python-browser.sh  run-python-wasi.sh  test-wasi.sh
root@27c144c9747d:/python-wasm# mkdir -p builddir/build
root@27c144c9747d:/python-wasm# pushd builddir/build
/python-wasm/builddir/build /python-wasm
root@27c144c9747d:/python-wasm/builddir/build# ls ../../configure
ls: cannot access '../../configure': No such file or directory
root@27c144c9747d:/python-wasm/builddir/build# ls ../../cpython/configure
../../cpython/configure
```

Comment (erlend-aasland) [reply Tools/wasm/README.md]: I'm marking this as resolved, since the new path is correct ;)

**Comment by erlend-aasland** [comment]
> Leave the `config.cache` in the instructions. It is a massive time safer for everybody that wants to hack on WASM. `emconfigure configure` takes several minutes even on a fast machine.

Sure! I'll revert the last commit.

File: Tools/wasm/README.md
Comment (tiran) [inline Tools/wasm/README.md:38]: I see where the instructions can be misunderstood. The instructions assume that all commands are executed from the cpython checkout outside the container **and** from mounted cpython checkout inside the container. Depending on how you set up the container, you may have ``/python-wasm`` as initial working directory. In that case you have to ``cd /python-wasm/cpython`` first.

File: Tools/wasm/README.md
Comment (tiran) [inline Tools/wasm/README.md]:
```suggestion
podman run --rm -ti -v $(pwd):/python-wasm/cpython:Z -w /python-wasm/cpython quay.io/tiran/cpythonbuild:emsdk3
```

File: Tools/wasm/README.md
Comment (tiran) [inline Tools/wasm/README.md]:
```suggestion
docker run --rm -ti -v $(pwd):/python-wasm/cpython -w /python-wasm/cpython quay.io/tiran/cpythonbuild:emsdk3
```

File: Tools/wasm/README.md
Comment (tiran) [inline Tools/wasm/README.md]:
```suggestion
CONFIG_SITE=../../Tools/wasm/config.site-wasm32-emscripten \
  emconfigure ../../configure -C \
```

File: Tools/wasm/README.md
Comment (tiran) [inline Tools/wasm/README.md]:
```suggestion
    --build=$(../../config.guess) \
```

File: Tools/wasm/README.md
Comment (tiran) [inline Tools/wasm/README.md]:
```suggestion
CONFIG_SITE=../../Tools/wasm/config.site-wasm32-emscripten \
  emconfigure ../../configure -C \
```

File: Tools/wasm/README.md
Comment (tiran) [inline Tools/wasm/README.md]:
```suggestion
    --build=$(../../config.guess) \
```

**Review by tiran — CHANGES_REQUESTED** [review]

File: Tools/wasm/README.md
Comment (erlend-aasland) [reply Tools/wasm/README.md:38]: > The instructions assume that all commands are executed from the cpython checkout outside the container [...]

Now I'm confused. The build instructions are to be executed from _inside_ the container, right?

**Comment by erlend-aasland** [comment]
Thanks, Christian. Specifying working dir is a better approach.

File: Tools/wasm/README.md
Comment (tiran) [inline Tools/wasm/README.md]: The work is not lost when you exit the container. With the correct setup the work is inside the CPython checkout.

```suggestion
```

**Review by tiran — CHANGES_REQUESTED** [review]

File: Tools/wasm/README.md
Comment (erlend-aasland) [reply Tools/wasm/README.md]: We specify docker options that explicitly remove the container after exit. Should we change this, then?

Comment (erlend-aasland) [reply Tools/wasm/README.md]:
```
$ docker run --help | grep "\-\-rm"
      --rm                             Automatically remove the container when it exits
```

Comment (tiran) [reply Tools/wasm/README.md]: Yes, that is correct. ``--rm`` removes the container so we don't end up with tons of dead container instances. All build artifacts are persisted in the mounted directory.

Comment (erlend-aasland) [reply Tools/wasm/README.md]: Are there further information missing regarding what the "correct setup" is? :)

Comment (erlend-aasland) [reply Tools/wasm/README.md]: Is 7db408ceb6b2c0c9e7859ea1b63d1a3ca76693e8 acceptable?

Comment (tiran) [reply Tools/wasm/README.md]: > Are there further information missing regarding what the "correct setup" is? :)

You tell me :) Do the instructions work for you? 

I have run the containers hundreds times. The commands are embedded in my muscle memory. I didn't even notice that the instructions were missing the CWD because it was so obvious to me.

File: Tools/wasm/README.md
Comment (tiran) [inline Tools/wasm/README.md]: I suggest that you remove the paragraph. It is confusing and distracting. All instructions in the README operate on data in the mounted directory.

Comment (erlend-aasland) [reply Tools/wasm/README.md]:
```suggestion
```

File: Tools/wasm/README.md
Comment (erlend-aasland) [reply Tools/wasm/README.md]: I think we're good now :)

**Comment by erlend-aasland** [comment]
Thanks for all your help, Christian!

**Comment by erlend-aasland** [comment]
If you're fine with the last change, I'd like to land this, @tiran.

**Review by tiran — APPROVED** [review]

**Pull request merged**
